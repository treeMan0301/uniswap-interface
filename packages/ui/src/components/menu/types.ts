import { BaseSyntheticEvent } from 'react'
import { GeneratedIcon, IconProps } from 'ui/src/components/factories/createIcon'
import { TextProps } from 'ui/src/components/text'
import { TouchableAreaProps } from 'ui/src/components/touchable'

export type MenuContentItem = {
  label: string
  onPress: (e: BaseSyntheticEvent) => void
  textProps?: TextProps
  Icon?: GeneratedIcon | ((props: IconProps) => JSX.Element)
  iconProps?: IconProps
  destructive?: boolean
} & TouchableAreaProps
