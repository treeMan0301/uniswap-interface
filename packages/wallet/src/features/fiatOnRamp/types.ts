import { FiatPurchaseTransactionInfo, TransactionDetails } from 'wallet/src/features/transactions/types'

export type FiatOnRampTransactionDetails = TransactionDetails & {
  typeInfo: FiatPurchaseTransactionInfo
}

/** @ref https://dashboard.moonpay.com/api_reference/client_side_api#currencies */
export type MoonpayCurrency = {
  id: string
  type: 'crypto' | 'fiat'
  name?: string
  code: string
  supportsTestMode: boolean
  supportsLiveMode: boolean
  notAllowedUSStates: string[]
  isSupportedInUS: boolean
  metadata?: {
    contractAddress: string
    chainId: string
  }
}
