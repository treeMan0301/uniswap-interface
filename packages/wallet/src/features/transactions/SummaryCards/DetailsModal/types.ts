import {
  ConfirmedSwapTransactionInfo,
  ExactInputSwapTransactionInfo,
  ExactOutputSwapTransactionInfo,
  OnRampPurchaseInfo,
  OnRampTransferInfo,
  UnknownTransactionInfo,
  WrapTransactionInfo,
} from 'wallet/src/features/transactions/types'

export type SwapTypeTransactionInfo =
  | ExactInputSwapTransactionInfo
  | ExactOutputSwapTransactionInfo
  | ConfirmedSwapTransactionInfo

import {
  ApproveTransactionInfo,
  FiatPurchaseTransactionInfo,
  NFTApproveTransactionInfo,
  NFTMintTransactionInfo,
  NFTTradeTransactionInfo,
  ReceiveTokenTransactionInfo,
  SendTokenTransactionInfo,
  TransactionType,
  TransactionTypeInfo,
  WCConfirmInfo,
} from 'wallet/src/features/transactions/types'

export function isApproveTransactionInfo(typeInfo: TransactionTypeInfo): typeInfo is ApproveTransactionInfo {
  return typeInfo.type === TransactionType.Approve
}

export function isFiatPurchaseTransactionInfo(typeInfo: TransactionTypeInfo): typeInfo is FiatPurchaseTransactionInfo {
  return typeInfo.type === TransactionType.FiatPurchase
}

export function isOnRampPurchaseTransactionInfo(typeInfo: TransactionTypeInfo): typeInfo is OnRampPurchaseInfo {
  return typeInfo.type === TransactionType.OnRampPurchase
}

export function isOnRampTransferTransactionInfo(typeInfo: TransactionTypeInfo): typeInfo is OnRampTransferInfo {
  return typeInfo.type === TransactionType.OnRampTransfer
}

export function isNFTApproveTransactionInfo(typeInfo: TransactionTypeInfo): typeInfo is NFTApproveTransactionInfo {
  return typeInfo.type === TransactionType.NFTApprove
}

export function isNFTMintTransactionInfo(typeInfo: TransactionTypeInfo): typeInfo is NFTMintTransactionInfo {
  return typeInfo.type === TransactionType.NFTMint
}

export function isNFTTradeTransactionInfo(typeInfo: TransactionTypeInfo): typeInfo is NFTTradeTransactionInfo {
  return typeInfo.type === TransactionType.NFTTrade
}

export function isReceiveTokenTransactionInfo(typeInfo: TransactionTypeInfo): typeInfo is ReceiveTokenTransactionInfo {
  return typeInfo.type === TransactionType.Receive
}

export function isSendTokenTransactionInfo(typeInfo: TransactionTypeInfo): typeInfo is SendTokenTransactionInfo {
  return typeInfo.type === TransactionType.Send
}

export function isSwapTransactionInfo(typeInfo: TransactionTypeInfo): typeInfo is SwapTypeTransactionInfo {
  return typeInfo.type === TransactionType.Swap
}

export function isWCConfirmTransactionInfo(typeInfo: TransactionTypeInfo): typeInfo is WCConfirmInfo {
  return typeInfo.type === TransactionType.WCConfirm
}

export function isWrapTransactionInfo(typeInfo: TransactionTypeInfo): typeInfo is WrapTransactionInfo {
  return typeInfo.type === TransactionType.Wrap
}

export function isUnknownTransactionInfo(typeInfo: TransactionTypeInfo): typeInfo is UnknownTransactionInfo {
  return typeInfo.type === TransactionType.Unknown
}
