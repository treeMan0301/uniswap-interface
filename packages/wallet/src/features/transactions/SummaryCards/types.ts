import { AuthTrigger } from 'uniswap/src/features/auth/types'
import { TransactionDetails } from 'uniswap/src/features/transactions/types/transactionDetails'
import { TransactionState } from 'uniswap/src/features/transactions/types/transactionState'
import { WalletChainId } from 'uniswap/src/types/chains'

export interface TransactionSummaryLayoutProps {
  authTrigger?: AuthTrigger
  transaction: TransactionDetails
  title?: string
  caption: string | JSX.Element
  icon?: JSX.Element
  index?: number
  onRetry?: () => void
}

export interface SummaryItemProps {
  authTrigger?: AuthTrigger
  transaction: TransactionDetails
  swapCallbacks?: SwapSummaryCallbacks
  index?: number
}

export interface SwapSummaryCallbacks {
  useLatestSwapTransaction: (address: string) => TransactionDetails | undefined
  useSwapFormTransactionState: (
    address: Address | undefined,
    chainId: WalletChainId | undefined,
    txId: string | undefined,
  ) => TransactionState | undefined
  onRetryGenerator?: (swapFormState: TransactionState | undefined) => () => void
}
