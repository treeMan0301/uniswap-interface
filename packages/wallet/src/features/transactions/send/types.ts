import { Currency, CurrencyAmount } from '@uniswap/sdk-core'
import { AssetType, NFTAssetType } from 'uniswap/src/entities/assets'
import { WalletChainId } from 'uniswap/src/types/chains'
import { Account } from 'wallet/src/features/wallet/accounts/types'

interface BaseSendParams {
  type: AssetType
  txId?: string
  account: Account
  chainId: WalletChainId
  toAddress: Address
  tokenAddress: Address
  currencyAmountUSD?: Maybe<CurrencyAmount<Currency>> // for analytics
}

export interface SendCurrencyParams extends BaseSendParams {
  type: AssetType.Currency
  amountInWei: string
}

export interface SendNFTParams extends BaseSendParams {
  type: NFTAssetType
  tokenId: string
}

export type SendTokenParams = SendCurrencyParams | SendNFTParams
