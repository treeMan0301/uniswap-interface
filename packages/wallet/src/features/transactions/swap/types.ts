import { Currency, CurrencyAmount } from '@uniswap/sdk-core'
import { CurrencyInfo } from 'uniswap/src/features/dataApi/types'
import { CurrencyField } from 'uniswap/src/features/transactions/transactionState/types'
import { WalletChainId } from 'uniswap/src/types/chains'
import { useTrade } from 'wallet/src/features/transactions/swap/trade/api/hooks/useTrade'
import { BaseDerivedInfo } from 'wallet/src/features/transactions/transfer/types'
import { WrapType } from 'wallet/src/features/transactions/types'

export type DerivedSwapInfo<
  TInput = CurrencyInfo,
  TOutput extends CurrencyInfo = CurrencyInfo,
> = BaseDerivedInfo<TInput> & {
  chainId: WalletChainId
  currencies: BaseDerivedInfo<TInput>['currencies'] & {
    [CurrencyField.OUTPUT]: Maybe<TOutput>
  }
  currencyAmounts: BaseDerivedInfo<TInput>['currencyAmounts'] & {
    [CurrencyField.OUTPUT]: Maybe<CurrencyAmount<Currency>>
  }
  currencyAmountsUSDValue: {
    [CurrencyField.INPUT]: Maybe<CurrencyAmount<Currency>>
    [CurrencyField.OUTPUT]: Maybe<CurrencyAmount<Currency>>
  }
  currencyBalances: BaseDerivedInfo<TInput>['currencyBalances'] & {
    [CurrencyField.OUTPUT]: Maybe<CurrencyAmount<Currency>>
  }
  focusOnCurrencyField: CurrencyField | null
  trade: ReturnType<typeof useTrade>
  wrapType: WrapType
  selectingCurrencyField?: CurrencyField
  txId?: string
  autoSlippageTolerance?: number
  customSlippageTolerance?: number
}
