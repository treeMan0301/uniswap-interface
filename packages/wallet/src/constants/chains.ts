// Based on https://github.com/Uniswap/interface/blob/main/src/constants/chains.ts

import { ImageSourcePropType } from 'react-native'
import { GeneratedIcon, Logos } from 'ui/src'
import {
  ARBITRUM_LOGO,
  BASE_LOGO,
  BLAST_LOGO,
  BNB_LOGO,
  ETHEREUM_LOGO,
  GOERLI_LOGO,
  MUMBAI_LOGO,
  OPTIMISM_LOGO,
  POLYGON_LOGO,
  ZORA_LOGO,
} from 'ui/src/assets'
import { config } from 'uniswap/src/config'
import { ChainId } from 'uniswap/src/types/chains'

/** Address that represents native currencies on ETH, Arbitrum, etc. */
export const DEFAULT_NATIVE_ADDRESS = '0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee'

export enum RPCType {
  Public = 'public',
  Private = 'private',
  PublicAlt = 'public_alternative',
}

export const ALL_SUPPORTED_CHAINS: string[] = Object.values(ChainId).map((c) => c.toString())

// DON'T CHANGE - order here determines ordering of networks in app
// TODO: [MOB-250] Add back in testnets once our endpoints support them
export const ALL_SUPPORTED_CHAIN_IDS: ChainId[] = [
  ChainId.Mainnet,
  ChainId.Polygon,
  ChainId.ArbitrumOne,
  ChainId.Optimism,
  ChainId.Base,
  ChainId.Bnb,
  ChainId.Blast,
]

export const TESTNET_CHAIN_IDS = [ChainId.Goerli, ChainId.PolygonMumbai]

export const ETHEREUM_CHAIN_IDS = [ChainId.Mainnet, ChainId.Goerli] as const

// Renamed from SupportedL1ChainId in web app
export type EthereumChainId = (typeof ETHEREUM_CHAIN_IDS)[number]

export const L2_CHAIN_IDS = [
  ChainId.ArbitrumOne,
  ChainId.Base,
  ChainId.Optimism,
  ChainId.Polygon,
  ChainId.PolygonMumbai,
  ChainId.Bnb,
  ChainId.Blast,
  ChainId.Zora,
] as const

// Renamed from SupportedL2ChainId in web app
export type L2ChainId = (typeof L2_CHAIN_IDS)[number]

export interface L1ChainInfo {
  readonly blockWaitMsBeforeWarning?: number
  readonly bridge?: string
  readonly docs: string
  readonly explorer: {
    name: string
    url: string
    logoLight: GeneratedIcon
    logoDark: GeneratedIcon
  }
  readonly infoLink: string
  readonly label: string
  readonly logoUrl?: string
  readonly logo?: ImageSourcePropType
  readonly rpcUrls?: Partial<{ [key in keyof RPCType as RPCType]: string }>
  readonly nativeCurrency: {
    name: string // 'Goerli ETH',
    symbol: string // 'gorETH',
    decimals: number // 18,
    address: string // '0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee'
    explorerLink?: string // Special override for native ETH explorer link
  }
  readonly wrappedNativeCurrency: {
    name: string // 'Wrapped Ether',
    symbol: string // 'WETH',
    decimals: number // 18,
    address: string // '0xb4fbf271143f4fbf7b91a5ded31805e42b2208d6'
  }
}
export interface L2ChainInfo extends L1ChainInfo {
  readonly bridge: string
  readonly statusPage?: string
}

export type ChainInfo = {
  readonly [chainId in L2ChainId]: L2ChainInfo
} & { readonly [chainId in EthereumChainId]: L1ChainInfo }

export const CHAIN_INFO: ChainInfo = {
  [ChainId.ArbitrumOne]: {
    blockWaitMsBeforeWarning: 600000, // 10 minutes
    bridge: 'https://bridge.arbitrum.io/',
    docs: 'https://offchainlabs.com/',
    explorer: {
      name: 'Arbiscan',
      url: 'https://arbiscan.io/',
      logoLight: Logos.ArbiscanLogoLight,
      logoDark: Logos.ArbiscanLogoDark,
    },
    infoLink: 'https://app.uniswap.org/explore/tokens/arbitrum',
    label: 'Arbitrum',
    logo: ARBITRUM_LOGO,
    nativeCurrency: {
      name: 'Arbitrum ETH',
      symbol: 'ETH',
      decimals: 18,
      address: DEFAULT_NATIVE_ADDRESS,
      explorerLink: 'https://arbiscan.io/chart/etherprice',
    },
    wrappedNativeCurrency: {
      name: 'Wrapped Ether',
      symbol: 'WETH',
      decimals: 18,
      address: '0x82af49447d8a07e3bd95bd0d56f35241523fbab1',
    },
    rpcUrls: { [RPCType.PublicAlt]: 'https://arb1.arbitrum.io/rpc' },
  },
  [ChainId.Mainnet]: {
    blockWaitMsBeforeWarning: 60000, // 1 minute
    docs: 'https://docs.uniswap.org/',
    explorer: {
      name: 'Etherscan',
      url: 'https://etherscan.io/',
      logoLight: Logos.EtherscanLogoLight,
      logoDark: Logos.EtherscanLogoDark,
    },
    infoLink: 'https://app.uniswap.org/explore',
    label: 'Ethereum',
    logo: ETHEREUM_LOGO,
    nativeCurrency: {
      name: 'Ethereum',
      symbol: 'ETH',
      decimals: 18,
      address: DEFAULT_NATIVE_ADDRESS,
      explorerLink: 'https://etherscan.io/chart/etherprice',
    },
    wrappedNativeCurrency: {
      name: 'Wrapped Ether',
      symbol: 'WETH',
      decimals: 18,
      address: '0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2',
    },
    rpcUrls: { [RPCType.Private]: 'https://rpc.mevblocker.io/?referrer=uniswapwallet' },
  },
  [ChainId.Goerli]: {
    blockWaitMsBeforeWarning: 180000, // 3 minutes
    docs: 'https://docs.uniswap.org/',
    explorer: {
      name: 'Etherscan',
      url: 'https://goerli.etherscan.io/',
      logoLight: Logos.EtherscanLogoLight,
      logoDark: Logos.EtherscanLogoDark,
    },
    infoLink: 'https://app.uniswap.org/explore',
    label: 'Görli',
    logo: GOERLI_LOGO,
    nativeCurrency: {
      name: 'Görli ETH',
      symbol: 'görETH',
      decimals: 18,
      address: DEFAULT_NATIVE_ADDRESS,
      explorerLink: 'https://etherscan.io/chart/etherprice', // goerli.etherscan.io doesn't work
    },
    wrappedNativeCurrency: {
      name: 'Wrapped Ether',
      symbol: 'WETH',
      decimals: 18,
      address: '0xb4fbf271143f4fbf7b91a5ded31805e42b2208d6',
    },
  },
  [ChainId.Base]: {
    blockWaitMsBeforeWarning: 600000,
    bridge: 'https://bridge.base.org/',
    docs: 'https://base.org/',
    explorer: {
      name: 'BaseScan',
      url: 'https://basescan.org/',
      logoLight: Logos.EtherscanLogoLight,
      logoDark: Logos.EtherscanLogoDark,
    },
    infoLink: 'https://app.uniswap.org/explore/tokens/base',
    label: 'Base',
    logo: BASE_LOGO,
    nativeCurrency: {
      name: 'Base ETH',
      symbol: 'ETH',
      decimals: 18,
      address: DEFAULT_NATIVE_ADDRESS,
      explorerLink: 'https://basescan.org/chart/etherprice',
    },
    wrappedNativeCurrency: {
      name: 'Wrapped Ether',
      symbol: 'WETH',
      decimals: 18,
      address: '0x4200000000000000000000000000000000000006',
    },
    rpcUrls: { [RPCType.Public]: 'https://mainnet.base.org' },
  },
  [ChainId.Bnb]: {
    blockWaitMsBeforeWarning: 600000,
    bridge: 'https://www.bnbchain.org/bridge',
    docs: 'https://www.bnbchain.org/',
    explorer: {
      name: 'BscScan',
      url: 'https://bscscan.com/',
      logoLight: Logos.EtherscanLogoLight,
      logoDark: Logos.EtherscanLogoDark,
    },
    infoLink: 'https://app.uniswap.org/explore/tokens/bnb',
    label: 'BNB',
    logo: BNB_LOGO,
    nativeCurrency: {
      name: 'Binance Coin',
      symbol: 'BNB',
      decimals: 18,
      address: '0xb8c77482e45f1f44de1745f52c74426c631bdd52',
    },
    wrappedNativeCurrency: {
      name: 'Wrapped BNB',
      symbol: 'WBNB',
      decimals: 18,
      address: '0xbb4cdb9cbd36b01bd1cbaebf2de08d9173bc095c',
    },
    rpcUrls: { [RPCType.Public]: config.quicknodeBnbRpcUrl },
  },
  [ChainId.Optimism]: {
    blockWaitMsBeforeWarning: 1200000, // 20 minutes
    bridge: 'https://gateway.optimism.io/',
    docs: 'https://optimism.io/',
    explorer: {
      name: 'OP Etherscan',
      url: 'https://optimistic.etherscan.io/',
      logoLight: Logos.OpEtherscanLogoLight,
      logoDark: Logos.OpEtherscanLogoDark,
    },
    infoLink: 'https://app.uniswap.org/explore/tokens/optimism',
    label: 'Optimism',
    logo: OPTIMISM_LOGO,
    nativeCurrency: {
      name: 'Optimistic ETH',
      symbol: 'ETH',
      decimals: 18,
      address: DEFAULT_NATIVE_ADDRESS,
      explorerLink: 'https://optimistic.etherscan.io/chart/etherprice',
    },
    wrappedNativeCurrency: {
      name: 'Wrapped Ether',
      symbol: 'WETH',
      decimals: 18,
      address: '0x4200000000000000000000000000000000000006',
    },
    rpcUrls: { [RPCType.PublicAlt]: 'https://mainnet.optimism.io' },
    statusPage: 'https://optimism.io/status',
  },
  [ChainId.Polygon]: {
    blockWaitMsBeforeWarning: 600000, // 10 minutes
    bridge: 'https://wallet.polygon.technology/bridge',
    docs: 'https://polygon.io/',
    explorer: {
      name: 'PolygonScan',
      url: 'https://polygonscan.com/',
      logoLight: Logos.PolygonscanLogoLight,
      logoDark: Logos.PolygonscanLogoDark,
    },
    infoLink: 'https://app.uniswap.org/explore/tokens/polygon',
    label: 'Polygon',
    logo: POLYGON_LOGO,
    nativeCurrency: {
      name: 'Polygon Matic',
      symbol: 'MATIC',
      decimals: 18,
      address: '0x0000000000000000000000000000000000001010',
    },
    wrappedNativeCurrency: {
      name: 'Wrapped MATIC',
      symbol: 'WMATIC',
      decimals: 18,
      address: '0x0d500b1d8e8ef31e21c99d1db9a6444d3adf1270',
    },
    rpcUrls: { [RPCType.PublicAlt]: 'https://polygon-rpc.com/' },
  },
  [ChainId.PolygonMumbai]: {
    blockWaitMsBeforeWarning: 600000, // 10 minutes
    bridge: 'https://wallet.polygon.technology/bridge',
    docs: 'https://polygon.io/',
    explorer: {
      name: 'PolygonScan',
      url: 'https://mumbai.polygonscan.com/',
      logoLight: Logos.PolygonscanLogoLight,
      logoDark: Logos.PolygonscanLogoDark,
    },
    infoLink: 'https://app.uniswap.org/explore/tokens/polygon',
    label: 'Polygon Mumbai',
    logo: MUMBAI_LOGO,
    nativeCurrency: {
      name: 'Polygon Mumbai Matic',
      symbol: 'mMATIC',
      decimals: 18,
      address: '0x0000000000000000000000000000000000001010',
    },
    wrappedNativeCurrency: {
      name: 'Wrapped MATIC',
      symbol: 'WMATIC',
      decimals: 18,
      address: '0x9c3c9283d3e44854697cd22d3faa240cfb032889',
    },
    rpcUrls: { [RPCType.PublicAlt]: 'https://rpc-endpoints.superfluid.dev/mumbai' },
  },
  [ChainId.Blast]: {
    blockWaitMsBeforeWarning: 600000,
    bridge: 'https://blast.io/en/bridge/',
    docs: 'https://docs.blast.io/about-blast/',
    explorer: {
      name: 'BlastScan',
      url: 'https://blastscan.io/',
      logoLight: Logos.EtherscanLogoLight,
      logoDark: Logos.EtherscanLogoDark,
    },
    infoLink: 'https://app.uniswap.org/explore/tokens/blast',
    label: 'Blast',
    logo: BLAST_LOGO,
    nativeCurrency: {
      name: 'Blast ETH',
      symbol: 'ETH',
      decimals: 18,
      address: DEFAULT_NATIVE_ADDRESS,
    },
    wrappedNativeCurrency: {
      name: 'Wrapped Ether',
      symbol: 'WETH',
      decimals: 18,
      address: '0x4300000000000000000000000000000000000004',
    },
    rpcUrls: { [RPCType.Public]: 'https://rpc.blast.io' },
  },
  [ChainId.Zora]: {
    blockWaitMsBeforeWarning: 600000,
    bridge: 'https://bridge.zora.energy/',
    docs: 'https://docs.zora.co/',
    explorer: {
      name: 'Zora explorer',
      url: 'https://explorer.zora.energy/',
      logoLight: Logos.EtherscanLogoLight,
      logoDark: Logos.EtherscanLogoDark,
    },
    infoLink: 'https://app.uniswap.org/explore/tokens/zora?chain=zora',
    label: 'Zora',
    logo: ZORA_LOGO,
    nativeCurrency: {
      name: 'Zora ETH',
      symbol: 'ETH',
      decimals: 18,
      address: DEFAULT_NATIVE_ADDRESS,
    },
    wrappedNativeCurrency: {
      name: 'Wrapped Ether',
      symbol: 'WETH',
      decimals: 18,
      address: '0x4200000000000000000000000000000000000006',
    },
    rpcUrls: { [RPCType.Public]: 'https://rpc.zora.energy/' },
  },
}
