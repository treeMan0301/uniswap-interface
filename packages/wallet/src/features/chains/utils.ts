import { BigNumber, BigNumberish } from 'ethers'
import { L2ChainId, L2_CHAIN_IDS } from 'uniswap/src/constants/chains'
import { Chain } from 'uniswap/src/data/graphql/uniswap-data-api/__generated__/types-and-hooks'
import {
  UniverseChainId,
  WALLET_SUPPORTED_CHAIN_IDS,
  WalletChainId,
} from 'uniswap/src/types/chains'
import { isJestRun } from 'utilities/src/environment'
import { PollingInterval } from 'wallet/src/constants/misc'

// Some code from the web app uses chainId types as numbers
// This validates them as coerces into SupportedChainId
export function toSupportedChainId(chainId?: BigNumberish): WalletChainId | null {
  // Support Goerli for testing
  const ids = isJestRun
    ? [UniverseChainId.Goerli, ...WALLET_SUPPORTED_CHAIN_IDS]
    : WALLET_SUPPORTED_CHAIN_IDS

  if (!chainId || !ids.map((c) => c.toString()).includes(chainId.toString())) {
    return null
  }
  return parseInt(chainId.toString(), 10) as WalletChainId
}

export function chainIdToHexadecimalString(chainId: WalletChainId): string {
  return BigNumber.from(chainId).toHexString()
}

export function hexadecimalStringToInt(hex: string): number {
  return parseInt(hex, 16)
}

export const isL2Chain = (chainId?: UniverseChainId): boolean =>
  Boolean(chainId && L2_CHAIN_IDS.includes(chainId as L2ChainId))

export function fromGraphQLChain(chain: Chain | undefined): WalletChainId | null {
  switch (chain) {
    case Chain.Ethereum:
      return UniverseChainId.Mainnet
    case Chain.Arbitrum:
      return UniverseChainId.ArbitrumOne
    case Chain.EthereumGoerli:
      return UniverseChainId.Goerli
    case Chain.Optimism:
      return UniverseChainId.Optimism
    case Chain.Polygon:
      return UniverseChainId.Polygon
    case Chain.Base:
      return UniverseChainId.Base
    case Chain.Bnb:
      return UniverseChainId.Bnb
    case Chain.Blast:
      return UniverseChainId.Blast
    case Chain.Avalanche:
      return UniverseChainId.Avalanche
    case Chain.Celo:
      return UniverseChainId.Celo
    case Chain.Zora:
      return UniverseChainId.Zora
    case Chain.Zksync:
      return UniverseChainId.Zksync
  }

  return null
}

export function getPollingIntervalByBlocktime(chainId?: WalletChainId): PollingInterval {
  return isL2Chain(chainId) ? PollingInterval.LightningMcQueen : PollingInterval.Fast
}

export function fromMoonpayNetwork(moonpayNetwork: string | undefined): WalletChainId | undefined {
  switch (moonpayNetwork) {
    case Chain.Arbitrum.toLowerCase():
      return UniverseChainId.ArbitrumOne
    case Chain.Optimism.toLowerCase():
      return UniverseChainId.Optimism
    case Chain.Polygon.toLowerCase():
      return UniverseChainId.Polygon
    case Chain.Bnb.toLowerCase():
      return UniverseChainId.Bnb
    // Moonpay still refers to BNB chain as BSC so including both BNB and BSC cases
    case 'bsc':
      return UniverseChainId.Bnb
    case Chain.Base.toLowerCase():
      return UniverseChainId.Base
    case Chain.Avalanche.toLowerCase():
      return UniverseChainId.Avalanche
    case Chain.Celo.toLowerCase():
      return UniverseChainId.Celo
    case undefined:
      return UniverseChainId.Mainnet
    default:
      return undefined
  }
}

export function fromUniswapWebAppLink(network: string | null): WalletChainId | null {
  switch (network) {
    case Chain.Ethereum.toLowerCase():
      return UniverseChainId.Mainnet
    case Chain.Arbitrum.toLowerCase():
      return UniverseChainId.ArbitrumOne
    case Chain.Optimism.toLowerCase():
      return UniverseChainId.Optimism
    case Chain.Polygon.toLowerCase():
      return UniverseChainId.Polygon
    case Chain.Base.toLowerCase():
      return UniverseChainId.Base
    case Chain.Bnb.toLowerCase():
      return UniverseChainId.Bnb
    case Chain.Blast.toLowerCase():
      return UniverseChainId.Blast
    case Chain.Avalanche.toLowerCase():
      return UniverseChainId.Avalanche
    case Chain.Celo.toLowerCase():
      return UniverseChainId.Celo
    case Chain.Zora.toLowerCase():
      return UniverseChainId.Zora
    case Chain.Zksync.toLowerCase():
      return UniverseChainId.Zksync
    default:
      throw new Error(`Network "${network}" can not be mapped`)
  }
}

export function toUniswapWebAppLink(chainId: WalletChainId): string | null {
  switch (chainId) {
    case UniverseChainId.Mainnet:
      return Chain.Ethereum.toLowerCase()
    case UniverseChainId.ArbitrumOne:
      return Chain.Arbitrum.toLowerCase()
    case UniverseChainId.Optimism:
      return Chain.Optimism.toLowerCase()
    case UniverseChainId.Polygon:
      return Chain.Polygon.toLowerCase()
    case UniverseChainId.Base:
      return Chain.Base.toLowerCase()
    case UniverseChainId.Bnb:
      return Chain.Bnb.toLowerCase()
    case UniverseChainId.Blast:
      return Chain.Blast.toLowerCase()
    case UniverseChainId.Avalanche:
      return Chain.Avalanche.toLowerCase()
    case UniverseChainId.Celo:
      return Chain.Celo.toLowerCase()
    case UniverseChainId.Zora:
      return Chain.Zora.toLowerCase()
    case UniverseChainId.Zksync:
      return Chain.Zksync.toLowerCase()
    default:
      throw new Error(`ChainID "${chainId}" can not be mapped`)
  }
}
