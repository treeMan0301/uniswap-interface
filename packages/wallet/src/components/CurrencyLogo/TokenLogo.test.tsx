import { ChainId } from 'uniswap/src/types/chains'
import { render } from 'wallet/src/test/test-utils'
import { TokenLogo } from './TokenLogo'

describe('TokenLogo', () => {
  it('renders without error', () => {
    const tree = render(
      <TokenLogo chainId={ChainId.ArbitrumOne} symbol="DAI" url="https://example.com" />
    )

    expect(tree).toMatchSnapshot()
  })

  describe('token image', () => {
    it('renders svg when url is svg', () => {
      const { queryByTestId } = render(
        <TokenLogo chainId={ChainId.ArbitrumOne} symbol="DAI" url="https://example.com/image.svg" />
      )

      const tokenRemoteSvg = queryByTestId('token-remote-svg')
      const tokenImage = queryByTestId('token-image')

      expect(tokenRemoteSvg).toBeTruthy()
      expect(tokenImage).toBeFalsy()
    })

    it('renders image when url is valid and not svg', () => {
      const { queryByTestId } = render(
        <TokenLogo chainId={ChainId.ArbitrumOne} symbol="DAI" url="https://example.com/image.png" />
      )

      const tokenRemoteSvg = queryByTestId('token-remote-svg')
      const tokenImage = queryByTestId('token-image')

      expect(tokenRemoteSvg).toBeFalsy()
      expect(tokenImage).toBeTruthy()
    })

    it('renders fallback text when url is not specified', () => {
      const { queryByText } = render(<TokenLogo chainId={ChainId.ArbitrumOne} symbol="DAI" />)

      const fallbackText = queryByText('DAI')

      expect(fallbackText).toBeTruthy()
    })

    it('renders fallback text when url is invalid', () => {
      const { queryByText } = render(
        <TokenLogo chainId={ChainId.ArbitrumOne} symbol="DAI" url="invalid-url" />
      )

      const fallbackText = queryByText('DAI')

      expect(fallbackText).toBeTruthy()
    })

    it('does not render fallback text when url is valid', () => {
      const { queryByText } = render(
        <TokenLogo chainId={ChainId.ArbitrumOne} symbol="DAI" url="https://example.com/image.svg" />
      )

      const fallbackText = queryByText('DAI')

      expect(fallbackText).toBeFalsy()
    })
  })

  describe('network logo', () => {
    it('renders network logo by default', () => {
      const { queryByTestId } = render(
        <TokenLogo chainId={ChainId.ArbitrumOne} symbol="DAI" url="https://example.com" />
      )

      const networkLogo = queryByTestId('network-logo')

      expect(networkLogo).toBeTruthy()
    })

    it('renders network logo when hideNetworkLogo is false', () => {
      const { queryByTestId } = render(
        <TokenLogo
          chainId={ChainId.ArbitrumOne}
          hideNetworkLogo={false}
          symbol="DAI"
          url="https://example.com"
        />
      )

      const networkLogo = queryByTestId('network-logo')

      expect(networkLogo).toBeTruthy()
    })

    it('does not render network logo when hideNetworkLogo is true', () => {
      const { queryByTestId } = render(
        <TokenLogo
          hideNetworkLogo
          chainId={ChainId.ArbitrumOne}
          symbol="DAI"
          url="https://example.com"
        />
      )

      const networkLogo = queryByTestId('network-logo')

      expect(networkLogo).toBeFalsy()
    })

    it('does not render network logo when chainId is not specified', () => {
      const { queryByTestId } = render(<TokenLogo symbol="DAI" url="https://example.com" />)

      const networkLogo = queryByTestId('network-logo')

      expect(networkLogo).toBeFalsy()
    })

    it('does not render network logo when chainId is Mainnet', () => {
      const { queryByTestId } = render(
        <TokenLogo chainId={ChainId.Mainnet} symbol="DAI" url="https://example.com" />
      )

      const networkLogo = queryByTestId('network-logo')

      expect(networkLogo).toBeFalsy()
    })
  })
})
