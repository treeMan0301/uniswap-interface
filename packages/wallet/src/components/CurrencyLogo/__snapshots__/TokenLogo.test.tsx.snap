// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TokenLogo renders without error 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "flexDirection": "column",
      "height": 40,
      "justifyContent": "center",
      "width": 40,
    }
  }
  testID="token-logo"
>
  <Image
    resizeMode="contain"
    source={
      {
        "uri": "https://example.com",
      }
    }
    style={
      [
        {
          "backgroundColor": "#2222220D",
          "borderColor": "#2222220D",
          "borderRadius": 20,
          "borderWidth": 0.5,
          "height": 40,
          "width": 40,
        },
      ]
    }
    testID="token-image"
  />
  <View
    style={
      {
        "borderBottomColor": "#FFFFFF",
        "borderBottomLeftRadius": 8,
        "borderBottomRightRadius": 8,
        "borderBottomWidth": 2,
        "borderLeftColor": "#FFFFFF",
        "borderLeftWidth": 2,
        "borderRightColor": "#FFFFFF",
        "borderRightWidth": 2,
        "borderStyle": "solid",
        "borderTopColor": "#FFFFFF",
        "borderTopLeftRadius": 8,
        "borderTopRightRadius": 8,
        "borderTopWidth": 2,
        "bottom": -2,
        "flexDirection": "column",
        "position": "absolute",
        "right": -3,
      }
    }
  >
    <View
      style={
        {
          "borderColor": "#FFFFFF",
          "borderRadius": 6,
          "flexDirection": "column",
          "overflow": "hidden",
        }
      }
      testID="network-logo"
    >
      <Image
        resizeMode="contain"
        source={
          {
            "process": [Function],
          }
        }
        style={
          {
            "height": 16,
            "width": 16,
          }
        }
      />
    </View>
  </View>
</View>
`;
